### PR TITLE
Fix index out of range when fail to get instance types

### LINKED ISSUE #6025: Autoscaler creashes with panic on EKS with arm64 node group
<!--
Please answer these questions before submitting your bug report. Thanks!
-->

**Which component are you using?**:
cluster-autocaler

<!--
Which autoscaling component hosted in this repository (cluster-autoscaler, vertical-pod-autoscaler, addon-resizer, helm charts) is the bug in?
-->

**What version of the component are you using?**:
Tried with 
* registry.k8s.io/autoscaling/cluster-autoscaler:1.25.0 
* registry.k8s.io/autoscaling/cluster-autoscaler:v1.27.3

<!--
What version of the relevant component are you using? Either the image tag or helm chart version.
-->

Component version:

**What k8s version are you using (`kubectl version`)?**:

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.2", GitCommit:"7f6f68fdabc4df88cfea2dcf9a19b2b830f1e647", GitTreeState:"clean", BuildDate:"2023-05-17T14:20:07Z", GoVersion:"go1.20.4", Compiler:"gc", Platform:"windows/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"26+", GitVersion:"v1.26.6-eks-a5565ad", GitCommit:"895ed80e0cdcca657e88e56c6ad64d4998118590", GitTreeState:"clean", BuildDate:"2023-06-16T17:34:03Z", GoVersion:"go1.19.10", Compiler:"gc", Platform:"linux/amd64"}
</pre></details>

**What environment is this in?**:
AWS EKS 1.26 with self managed nodepools, mixed arm64 and amd64 node groups

Requirements given will result empty set
![image](https://github.com/kubernetes/autoscaler/assets/2135514/ad333317-6551-45d4-b7b6-185fcaee8dbf)

<!--
If you're using a cloud provider or hardware configuration as your deployment environment let us know here.
-->

**What did you expect to happen?**:
cluster-autoscaler should log error and ignore ASG group.
<!--
What behaviour did you expect to see?
-->
cluster-autoscaler crashes with similar messaging:
```log
I0808 11:11:09.005121       1 aws_cloud_provider.go:409] Successfully load 625 EC2 Instance Types [g5.2xlarge m5d.24xlarge r6in.xlarge x2gd.metal c6gd.metal c5n.9xlarge g3.4xlarge i3en.xlarge r5ad.large c6 │
│ panic: runtime error: index out of range [0] with length 0                                                                                                                                                    │
│                                                                                                                                                                                                               │
│ goroutine 48 [running]:                                                                                                                                                                                       │
│ k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws.(*awsWrapper).getInstanceTypeFromInstanceRequirements(0xc0014ec9c0, {0xc0017f3e30, 0x15}, 0xc000b23c20)                                                │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/aws_wrapper.go:327 +0x569                                                                                                              │
│ k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws.(*awsWrapper).getInstanceTypeFromRequirementsOverrides(0xc000a9f468?, 0xc001806930)                                                                    │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/aws_wrapper.go:259 +0x8c                                                                                                               │
│ k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws.(*awsWrapper).getInstanceTypesForAsgs(0xc001501700?, {0xc000de0400, 0xc, 0x1000000000000?})                                                            │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/aws_wrapper.go:721 +0xcbe                                                                                                              │
│ k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws.instanceTypeExpirationStore.populate({{0x5634680, 0xc001501700}, {0x562dba8, 0xc0014ecae0}, 0xc0014ec9c0}, 0x1?)                                       │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/instance_type_cache.go:116 +0x278                                                                                                      │
│ k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws.(*asgCache).regenerate(0xc0007c99d0)                                                                                                                   │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go:408 +0x905                                                                                                      │
│ k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws.(*AwsManager).forceRefresh(0xc0014dae40)                                                                                                               │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/aws_manager.go:261 +0x3a                                                                                                               │
│ k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws.createAWSManagerInternal({0x0?, 0x0?}, {{0x0, 0x0, 0x0}, {0xc000b51d70, 0x1, 0x1}}, 0x0, 0xc000b99740)                                                 │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/aws_manager.go:224 +0x6d2                                                                                                              │
│ k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws.CreateAwsManager(...)                                                                                                                                  │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/aws_manager.go:248                                                                                                                     │
│ k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws.BuildAWS({{0x3fe0000000000000, 0x3fe0000000000000, 0x8bb2c97000, 0x1176592e000}, 0xa, 0x0, 0x4e200, 0x0, 0x186a0000000000, 0x0, ...}, ...)             │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go:412 +0x626                                                                                                       │
│ k8s.io/autoscaler/cluster-autoscaler/cloudprovider/builder.buildCloudProvider({{0x3fe0000000000000, 0x3fe0000000000000, 0x8bb2c97000, 0x1176592e000}, 0xa, 0x0, 0x4e200, 0x0, 0x186a0000000000, 0x0, ...}, .. │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/builder/builder_all.go:95 +0xf1                                                                                                            │
│ k8s.io/autoscaler/cluster-autoscaler/cloudprovider/builder.NewCloudProvider({{0x3fe0000000000000, 0x3fe0000000000000, 0x8bb2c97000, 0x1176592e000}, 0xa, 0x0, 0x4e200, 0x0, 0x186a0000000000, 0x0, ...})      │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/builder/cloud_provider_builder.go:45 +0x1ce                                                                                                │
│ k8s.io/autoscaler/cluster-autoscaler/core.initializeDefaultOptions(0xc000aa14e8)                                                                                                                              │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/autoscaler.go:104 +0x318                                                                                                                            │
│ k8s.io/autoscaler/cluster-autoscaler/core.NewAutoscaler({{{0x3fe0000000000000, 0x3fe0000000000000, 0x8bb2c97000, 0x1176592e000}, 0xa, 0x0, 0x4e200, 0x0, 0x186a0000000000, 0x0, ...}, ...})                   │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/autoscaler.go:67 +0x36                                                                                                                              │
│ main.buildAutoscaler({0x5631eb8, 0xc000a39b00})                                                                                                                                                               │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:390 +0x698                                                                                                                                       │
│ main.run(0x0?, {0x5631eb8, 0xc000a39b00})                                                                                                                                                                     │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:396 +0x65                                                                                                                                        │
│ main.main.func2({0x0?, 0xc000142660?})                                                                                                                                                                        │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:508 +0x25                                                                                                                                        │
│ created by k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run                                                                                                                                         │
│     /gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:211 +0x11b                                                                                │
│ Stream closed EOF for kube-system/cluster-autoscaler-aws-cluster-autoscaler-574c95bd8b-rmth9 (aws-cluster-autoscaler)
```
**What happened instead?**:
ASG 
<!--
What behaviour did see instead?
-->

**How to reproduce it (as minimally and precisely as possible)**:

<!--
If possible, provide a recipe for reproducing the error.
A detailed sequence of steps describing what to do to observe the issue is good.
A complete runnable bash shell script is best.
-->

**Anything else we need to know?**:

<!--
Is there anything else you think we should know? Configuration of the component (be careful what you post here if so)? Relevant logs?
-->

### PR DESCRIPTION
**Pull request by fgksgf**
#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:

Fixes #6025 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by gjtempleton** [comment]
Hey @fgksgf , thanks for the PR, do you think it's worth adding a test covering an empty list being returned from `GetInstanceTypesFromInstanceRequirements`?

**Comment by fgksgf** [comment]
> Hey @fgksgf , thanks for the PR, do you think it's worth adding a test covering an empty list being returned from `GetInstanceTypesFromInstanceRequirements`?

Yes, I'll add one, thank you.

File: cluster-autoscaler/cloudprovider/aws/aws_wrapper.go [hunk 1]
```diff
@@ -326,15 +326,15 @@ func (m *awsWrapper) getInstanceTypeFromInstanceRequirements(imageId string, req
...
+	var instanceTypes []string
 	err = m.GetInstanceTypesFromInstanceRequirementsPages(requirementsInput, func(page *ec2.GetInstanceTypesFromInstanceRequirementsOutput, isLastPage bool) bool {
 		for _, instanceType := range page.InstanceTypes {
 			instanceTypes = append(instanceTypes, *instanceType.InstanceType)
 		}
 		return !isLastPage
 	})
 	observeAWSRequest("GetInstanceTypesFromInstanceRequirements", err, start)
-	if err != nil {
+	if err != nil || len(instanceTypes) == 0 {
```
Comment (drmorr0) [inline cluster-autoscaler/cloudprovider/aws/aws_wrapper.go:337]: I wonder if it would be better to have two different conditionals here: one that checks if `err != nil`, and one that checks if the number of returned instance types is 0?  As it currently stands, someone reading the logs can't tell whether things failed because of, say, a network error versus "the cloud provider couldn't find any matching instances".

Bonus points if we wrap the returned error with `%w` in the `err != nil` case so we can see exactly what went wrong on the cloud provider's end.

**Review by drmorr0 — CHANGES_REQUESTED** [review]

File: cluster-autoscaler/cloudprovider/aws/aws_wrapper_test.go [hunk 2]
```diff
@@ -695,3 +695,42 @@ func TestBuildLaunchTemplateFromSpec(t *testing.T) {
...
+	).Run(func(args mock.Arguments) {
+		fn := args.Get(1).(func(*ec2.GetInstanceTypesFromInstanceRequirementsOutput, bool) bool)
+		fn(&ec2.GetInstanceTypesFromInstanceRequirementsOutput{
+			InstanceTypes: []*ec2.InstanceTypeInfoFromInstanceRequirements{},
+		}, false)
+	}).Return(nil)
+
+	result, err := awsWrapper.getInstanceTypeFromInstanceRequirements("123", requirements)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "unable to get instance types from requirements")
```
Comment (drmorr0) [inline cluster-autoscaler/cloudprovider/aws/aws_wrapper_test.go:734]: Can we use `assert.ErrorIs` here instead of `Contains`?

File: cluster-autoscaler/cloudprovider/aws/aws_wrapper.go (hunk 1 above)
Comment (fgksgf) [reply cluster-autoscaler/cloudprovider/aws/aws_wrapper.go:337]: make sense to me

File: cluster-autoscaler/cloudprovider/aws/aws_wrapper_test.go (hunk 2 above)
Comment (fgksgf) [reply cluster-autoscaler/cloudprovider/aws/aws_wrapper_test.go:734]: I just found `assert.EqualError` is more commonly used in the codebase :P
